backend/rates/coingecko.go

```diff
@@ -0,0 +1,153 @@
+package rates
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/digitalbitbox/bitbox-wallet-app/util/observable"
+	"github.com/digitalbitbox/bitbox-wallet-app/util/observable/action"
+)
+
+const coingeckoAPIV3 = "https://api.coingecko.com/api/v3"
+
+// forwardHistoryLoop periodically updates historical market exchange rates
+// forward in time starting with the last fetched timestamp.
+func (updater *RateUpdater) forwardHistoryLoop() {
+	for {
+		// When to update next, after this loop iteration is done.
+		untilNext := 10 * time.Minute // TODO: add jitter
+
+		start := updater.histLatestTimestamp("BTC")
+		if !start.IsZero() {
+			// Start slightly past the last fetched timestamp
+			start = start.Add(10 * time.Minute)
+			now := time.Now()
+			// It's ok if start == now or now < start. CoinGecko simply returns no results.
+			// It's also ok if our "now" doesn't match CoinGecko: it always returns
+			// the latest available data without errors.
+			err := updater.updateHist("BTC", "USD", start, now)
+			if err != nil {
+				updater.log.Printf("updateHist(%v, %v): %v", start, now, err)
+				untilNext = time.Minute // TODO: exponential backoff
+			}
+		}
+
+		time.Sleep(untilNext)
+	}
+}
+
+// backwardHistoryLoop updates historical market exchange rates starting with
+// the earliest fetched timestamp backwards in time.
+func (updater *RateUpdater) backwardHistoryLoop() {
+	for {
+		// When to update next, after this loop iteration is done.
+		untilNext := 10 * time.Minute // TODO: add jitter
+
+		end := updater.histEarliestTimestamp("BTC")
+		if end.IsZero() {
+			end = time.Now()
+		}
+		// 90 days is the max interval which CoinGecko responds with hourly timeseries.
+		start := end.Add(-90 * 24 * time.Hour)
+		if time.Since(start) > 720*24*time.Hour {
+			// ~2 years of historical data is enough for now.
+			// TODO: Where do we stop?
+			return
+		}
+
+		err := updater.updateHist("BTC", "USD", start, end)
+		if err != nil {
+			updater.log.Printf("updateHist(%s, %s): %v", start, end, err)
+			untilNext = time.Minute // TODO: exponential backoff
+		}
+
+		time.Sleep(untilNext)
+	}
+}
+
+func (updater *RateUpdater) updateHist(coin, fiat string, start, end time.Time) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+	rates, err := updater.fetchMarketRange(ctx, coin, fiat, start, end)
+	if err != nil {
+		return err
+	}
+
+	updater.histMu.Lock()
+	defer updater.histMu.Unlock()
+	// TODO: updater.hist[coin] and rates may have duplicate timestamps.
+	a := append(updater.hist[coin], rates...)
+	sort.Slice(updater.hist[coin], func(i, j int) bool {
+		return a[i].timestamp.Before(a[j].timestamp)
+	})
+	updater.hist[coin] = a
+
+	updater.Notify(observable.Event{
```

Comment:
i think this isn't how the events work, in .Reload the Data should be sent with it, and the subject should point to a valid handlers endpoint.

I would remove this for now, we can add it in a separate PR that deals with updating the chart.

Reply:
removed